### PR TITLE
Improvements for PKCS#11 library

### PR DESCRIPTION
**Pull request by SparkiDev**

### CODE REVIEW

File: wolfcrypt/src/wc_pkcs11.c [hunk 1]
```diff
@@ -947,6 +963,7 @@ static int Pkcs11RsaKeyGen(Pkcs11Session* session, wc_CryptoInfo* info)
         { CKA_VERIFY,          &ckTrue,  sizeof(ckTrue)  },
         { CKA_PUBLIC_EXPONENT, &pub_exp, sizeof(pub_exp) }
     };
+    int               pubTmplCnt = sizeof(pubKeyTmpl)/sizeof(*pubKeyTmpl);
```
Comment (dgarske) [inline wolfcrypt/src/wc_pkcs11.c]: You may get a cast warning here from size_t to int. Recommend adding `(int)` cast here.

File: wolfcrypt/src/wc_pkcs11.c [hunk 2]
```diff
@@ -148,25 +150,36 @@ void wc_Pkcs11_Finalize(Pkcs11Dev* dev)
...
 
     if (ret == 0) {
         if (slotId < 0) {
             /* Use first available slot with a token. */
-            rv = dev->func->C_GetSlotList(CK_TRUE, &slot, &slotCnt);
+            rv = dev->func->C_GetSlotList(CK_TRUE, NULL, &slotCnt);
             if (rv != CKR_OK)
                 ret = WC_HW_E;
+            if (ret == 0) {
+                slot = XMALLOC(slotCnt * sizeof(*slot), NULL,
```
Comment (dgarske) [inline wolfcrypt/src/wc_pkcs11.c]: Please add cast here `(CK_SLOT_ID*)`.

Comment (dgarske) [reply wolfcrypt/src/wc_pkcs11.c]: Do we have a dev->heap to use for the XMALLOC?

File: wolfcrypt/src/wc_pkcs11.c [hunk 3]
```diff
@@ -178,6 +191,9 @@ int wc_Pkcs11Token_Init(Pkcs11Token* token, Pkcs11Dev* dev, int slotId,
         token->userPinSz = (CK_ULONG)userPinSz;
     }
 
+    if (slot != NULL)
+        XFREE(slot, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (dgarske) [inline wolfcrypt/src/wc_pkcs11.c]: If heap hint is available please add here as well dev->heap.

File: wolfcrypt/src/wc_pkcs11.c [hunk 4]
```diff
@@ -148,25 +150,36 @@ void wc_Pkcs11_Finalize(Pkcs11Dev* dev)
...
 
     if (ret == 0) {
         if (slotId < 0) {
             /* Use first available slot with a token. */
-            rv = dev->func->C_GetSlotList(CK_TRUE, &slot, &slotCnt);
+            rv = dev->func->C_GetSlotList(CK_TRUE, NULL, &slotCnt);
             if (rv != CKR_OK)
                 ret = WC_HW_E;
+            if (ret == 0) {
+                slot = (CK_SLOT_ID*)XMALLOC(slotCnt * sizeof(*slot), NULL,
```
Comment (toddouska) [inline wolfcrypt/src/wc_pkcs11.c]: Let's add a dev->heap member that gets set during the device init() for all XMALLOCs and FREEs.   Much easier to do now and maintain then to try and add later with the interfaces already out there.

**Pull request merged**